File: C++/Newton-Raphson.cpp
// Código creado por Joshua Mijangos López, 02/10/24
// Este código está hecho para que las funciones del método se modifiquen desde aquí
// Este código no acepta de datos de ingreso desde consola, recomiendo ejecutarlo desde Powershell
// En algunas líneas utilizo "cout" y en otras "printf" debido a los caracteres especiales

#include <iostream>
#include <cmath>
#include <iomanip>

using namespace std;

// Definimos la función f(x) y su derivada f'(x)
double f(double x) {
    return pow(x, 3) - x - 1;  // f(x) = x^3 - x - 1
}

double df(double x) {
    return 3 * pow(x, 2) - 1;  // f'(x) = 3x^2 - 1
}

int main() {
    double x0 = 1;  // Aproximación inicial
    double tol = 1e-5;  // Tolerancia
    int nmax = 100;  // Número máximo de iteraciones
    double error = 100;
    int niter = 0;

    cout << "# de Iteraciones\t\t  x_n \t\t    f(x_n) \t\t   Error\n";

    // Método de Newton
    while (error > tol && niter < nmax) {
        double fx0 = f(x0);
        double dfx0 = df(x0);

        // Verificar si la derivada es cero
        if (dfx0 == 0) {
            cout << "La derivada es cero, no se puede aplicar el método de Newton." << endl;
            break;
        }

        double x1 = x0 - fx0 / dfx0;  // Fórmula de Newton

        // Primera iteración sin error
        if (niter == 0) {
            cout << "\t" << niter + 1 << "\t\t " << fixed << setprecision(6) << x0 << "\t\t " << fx0 << "\t\t   N/A\n";
        } else {
            error = fabs(x1 - x0);  // Calcula el error
            cout << "\t" << niter + 1 << "\t\t " << fixed << setprecision(6) << x0 << "\t\t " << fx0 << "\t\t " << error << "\n";
        }

        // Actualizar para la siguiente iteración
        x0 = x1;
        niter++;
    }

    // Imprimir resultado final
    if (error <= tol) {
        cout << "\nSe realizaron " << niter << " iteraciones.\n";
        printf("La ra%cz aproximada de la funci%cn f(x) es %.6f, con un error de %.5f\n", 161, 162, x0, tol);
        // El segundo parámetro %.zf debe de modificarse si lo que se busca es disminuir el error, lo dejé preestablecido para z=5 (Es decir, un error de 10^-5)
    } else {
        printf("El m%ctodo no convergi%c en el n%cmero m%cximo de iteraciones.\n", 130, 162, 163, 160 );
    }

    return 0;
}

File: Matlab/Newton-Raphson.m
% Código creado por Joshua Mijangos López, 02/10/24
% Este código está hecho para que las funciones del método se modifiquen desde aquí
% Este código no acepta de datos de ingreso desde consola

% Definimos la variable simbólica y la función f(x)
syms x;
f_sym = x^3 - x - 1;  % f(x) = x^3 - x - 1
df_sym = diff(f_sym, x);  % Derivada f'(x)

% Convertimos a funciones anónimas para evaluación numérica
f = matlabFunction(f_sym);
df = matlabFunction(df_sym);

% Imprimir las funciones en pantalla
fprintf('Sean las funciones: \n\n');
disp(['f(x) = ', char(f_sym)]);
disp(['f''(x) = ', char(df_sym)]); % Usamos char para convertir la expresión simbólica a texto

fprintf('\n\n')

% Parámetros de entrada
x0 = 1;  % Aproximación inicial
tol = 1e-5;  % Tolerancia
nmax = 100;  % Número máximo de iteraciones
error = 100;
niter = 0;

fprintf('# de Iteración\t\tx_n\t\tf(x_n)\t\tError\n');

% Método de Newton
while error > tol && niter < nmax
    fx0 = f(x0);
    dfx0 = df(x0);
    
    if dfx0 == 0
        disp('La derivada es cero, no se puede aplicar el método de Newton.');
        break;
    end
    
    x1 = x0 - fx0 / dfx0;  % Fórmula del método de Newton
    
    if niter == 0
        fprintf('%d\t\t%.6f\t\t%.6f\t\tN/A\n', niter+1, x0, fx0);
    else
        error = abs(x1 - x0);  % Cálculo del error
        fprintf('%d\t\t%.6f\t\t%.6f\t\t%.6f\n', niter+1, x0, fx0, error);
    end
    
    x0 = x1;
    niter = niter + 1;
end

% Resultado final
if error <= tol
    fprintf('\nSe realizaron %d iteraciones.\n', niter);
    fprintf('La raíz aproximada de la función f(x) es %.6f con un error de %.5f\n', x0, tol);
    % El segundo parámetro %.zf debe de modificarse si lo que se busca es disminuir el error, lo dejé preestablecido para z=5 (Es decir, un error de 10^-5)
else
    disp('El método no convergió en el número máximo de iteraciones.');
end
